swap/src/bin/swap.rs

```diff
@@ -268,6 +273,86 @@ async fn main() -> Result<()> {
+                .context("Failed to read in seed file")?;
+
+            let identity = seed.derive_libp2p_identity();
+
+            // TODO: Compose behaviour that allows to requests Quotes as well
+            let behaviour = list_makers::Behaviour {
+                rendezvous: Rendezvous::new(identity, rendezvous::Config::default()),
+                quote: quote::cli(),
+            };
+            let mut swarm = swarm::cli(&seed, tor_socks5_port, behaviour).await?;
+
+            let _ = swarm.dial_addr(rendezvous_node_addr.clone());
+
+            while let Some(event) = swarm.next().await {
+                match event {
+                    SwarmEvent::ConnectionEstablished { peer_id, .. }
+                        if peer_id == rendezvous_node_peer_id =>
+                    {
+                        tracing::info!(
+                            "Connected to rendezvous point, discovering nodes in '{}' namespace ...",
+                            namespace
+                        );
+
+                        swarm.behaviour_mut().rendezvous.discover(
+                            Some(Namespace::new(namespace.clone()).unwrap()),
+                            None,
+                            None,
+                            rendezvous_node_peer_id,
+                        );
+                    }
+                    SwarmEvent::UnreachableAddr { error, address, .. }
+                    | SwarmEvent::UnknownPeerUnreachableAddr { error, address, .. }
+                        if address == rendezvous_node_addr =>
+                    {
+                        tracing::error!(
+                            "Failed to connect to rendezvous point at {}: {}",
+                            address,
+                            error
+                        );
+                        todo!("Error handling")
+                    }
+                    SwarmEvent::Behaviour(list_makers::OutEvent::Rendezvous(
+                        rendezvous::Event::Discovered { registrations, .. },
+                    )) => {
+                        for registration in registrations {
+                            for address in registration.record.addresses() {
+                                let peer = registration.record.peer_id();
+                                tracing::info!("Discovered peer {} at {}", peer, address);
+
+                                let p2p_suffix = Protocol::P2p(*peer.as_ref());
+                                let _address_with_p2p = if !address
+                                    .ends_with(&Multiaddr::empty().with(p2p_suffix.clone()))
+                                {
+                                    address.clone().with(p2p_suffix)
+                                } else {
+                                    address.clone()
+                                };
+
+                                todo!("Dial Alice and request quote!")
+                            }
+                        }
+                    }
```

Comment:
Some thoughts:

1. There could be more than one address per peer but we only want to display one quote.
2. We need to account for the fact that a connection might be unsuccessful, we shouldn't wait for that "forever". i.e. we can't just naively wait for 5 quote responses if we discovered 5 ASBs.

Reply:
I did not think about the multiple addresses that could be available for one peer. Thanks for mentioning that. 

I propose to store a relation of 

```
PeerId -> Multiaddress(es) -> Quote status
```

So upon discovery we trigger a quote request for each address (I think we must do that, because otherwise it gets complicated to decide when to trigger the next address...).
Once we get a quote we record it for the peer-id and mark it as done. If we record another quote coming in for the same peer we ignore it. 
If we run into an unreachable we store that for the specific address of the peer. If all addresses are marked as "request failed" we know that this peer was generally unreachable. 

So the break conditions for the loop are: 
For each peer, either quote received for peer (don't care about the address for now), or all quote requests for all addresses failed. 

I think this should be complete, so we don't need an additional timeout around the complete loop.

Reply:
> So upon discovery we trigger a quote request for each address (I think we must do that, because otherwise it gets complicated to decide when to trigger the next address...).

You will need to `dial` first, before you can request a quote. I think the best way to handle multiple addresses is to add all the addresses you discover to the quote protocol via `add_address`.

Then, you can send a quote request which will initiate a dial attempt (this is `libp2p-request-reponse` behaviour).

Eventually, you will either get one successful or failed event from `Quote` per Peer that you tried.